cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -345,6 +362,160 @@ public ApplicationWithPrograms updateApp(ApplicationId appId, AppRequest appRequ
                      programTerminator, ownerAdmin.getOwner(appId), appRequest.canUpdateSchedules());
   }
 
+  /**
+   * Upgrades an existing application by upgrading appliation artifact versions and plugin artifact versions.
+   *
+   * @param appId the id of the application to upgrade.
+   * @param programTerminator a program terminator that will stop programs that are removed when updating an app.
+   *                          For example, if an update removes a flow, the terminator defines how to stop that flow.
+   * @return Details of upgrade for the application like failure details with reason.
+   * @return
+   * @throws Exception
+   */
+  public ApplicationUpdateDetails upgradeApplication(ApplicationId appId, ProgramTerminator programTerminator)
+      throws Exception {
+    // Check if the current user has admin privileges on it before updating.
+    authorizationEnforcer.enforce(appId, authenticationContext.getPrincipal(), Action.ADMIN);
+    // check that app exists
+    ApplicationSpecification currentSpec = store.getApplication(appId);
+    if (currentSpec == null) {
+      LOG.warn("Application {} not found for upgrade.", appId);
+      return new ApplicationUpdateDetails(new NotFoundException(appId));
+    }
+    ArtifactId currentArtifact = currentSpec.getArtifactId();
+    NamespaceId currentArtifactNamespace =
+        ArtifactScope.SYSTEM.equals(currentArtifact.getScope()) ? NamespaceId.SYSTEM : appId.getParent();
```

Comment:
indentation should be 2 spaces. Same comment for rest of PR

Reply:
Sorry for that. I lost the intellij settings and things reset to previous state. Planning to set it up again. I tried to manually catch as much I could find for now. Looking for more.